File: tests/testthat/test-panzar_rosse.R
library(testthat)

test_that("Panzar-Rosse H-statistic matches expected per period", {
  df <- build_df()
  res_panzar <- MarketCompetitionMetrics$panzar_rosse(
    df,
    revenue_col = "Revenue",
    input_cols = c("Labor_cost", "Capital_cost"),
    period_col = "Period",
    plot = FALSE
  )

  for (p in names(expected_panzar)) {
    val <- extract_period_value(res_panzar, p, key_candidates = c("H","h","H_stat","H_statistic"))
    expect_false(is.na(val), info = paste("Panzar-Rosse H is NA for period", p))
    expect_equal(as.numeric(val), as.numeric(expected_panzar[p]), tolerance = 1e-6)
  }
})

File: tests/pytest/test_panzar_rosse.py
# tests/pytest/test_panzar_rosse.py
import numpy as np
import pytest
from tests.pytest._helpers import _build_df, _extract_period_value, _expected

@pytest.mark.parametrize("period,expected", list(_expected["panzar"].items()))
def test_panzar_rosse(period, expected):
    df = _build_df()
    from market_competition_metrics import MarketCompetitionMetrics

    res = MarketCompetitionMetrics.panzar_rosse(
        df,
        revenue_col="Revenue",
        input_cols=["Labor_cost", "Capital_cost"],
        period_col="Period",
        plot=False
    )

    value = _extract_period_value(res, period, key_candidates=["H","H_t","h","value"])
    assert np.isclose(value, expected, rtol=2e-3, atol=1e-6), f"Panzar-Rosse H mismatch for {period}: got {value}, expected {expected}"
